File: laravel/app/Models/Tournaments/TournamentRound.php
<?php

namespace App\Models\Tournaments;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;

class TournamentRound extends Model
{
    protected $table = 'tournament_rounds';

    protected $fillable = ['round_number', 'status', 'started_at', 'ended_at', 'time_limit_minutes', 'tournament_id'];

    protected $casts = [
        'started_at' => 'datetime',
        'ended_at' => 'datetime',
    ];

    const STATUS_VALUES = ['Pending', 'Active', 'Completed'];

    public function tournament(): BelongsTo
    {
        return $this->belongsTo(Tournament::class, 'tournament_id');
    }

    // ── Validation rules ─────────────────────────────────────────────
    public function validateRules(): void
    {
        $errors = [];
        if (!(($this->round_number === null || $this->round_number > 0))) {
            $errors['round_number_positive'] = 'Round number must be greater than zero';
        }
        if (!(($this->time_limit_minutes === null || $this->time_limit_minutes > 0))) {
            $errors['time_limit_positive'] = 'Round time limit must be greater than zero';
        }
        if (!empty($errors)) {
            throw new \Illuminate\Validation\ValidationException(
                \Illuminate\Support\Facades\Validator::make([], []),
                response()->json(['errors' => $errors], 422)
            );
        }
    }

    // ── Domain invariants (IMPLIES rules) ───────────────────────────────
    public function validateImplies(): void
    {
        if ($this->ended_at !== null && !(($this->ended_at === null || ($this->started_at !== null && $this->ended_at > $this->started_at)))) {
            throw new \RuntimeException('Round end time must be after start time');
        }
        if ($this->status === 'Completed' && $this->started_at === null) {
            throw new \RuntimeException('Completed round must have a start time');
        }
    }

    // ── Business operations ──────────────────────────────────────────

    public function start(): void
    {
        throw new \RuntimeException('start not implemented');
    }

    public function complete(): void
    {
        throw new \RuntimeException('complete not implemented');
    }

    public function generatePairings(): void
    {
        throw new \RuntimeException('generate_pairings not implemented');
    }

    public function isTimeExpired(): bool
    {
        throw new \RuntimeException('is_time_expired not implemented');
    }

}

File: phoenix/lib/cards_project/tournaments/tournament_round.ex
defmodule CardsProject.Tournaments.TournamentRound do
  use Ecto.Schema
  import Ecto.Changeset

  schema "tournament_rounds" do
    field :round_number, :integer
    field :status, :string
    field :started_at, :naive_datetime
    field :ended_at, :naive_datetime
    field :time_limit_minutes, :integer, default: 50
    belongs_to :tournament, CardsProject.Tournaments.Tournament
    belongs_to :matches, CardsProject.Tournaments.Match

    timestamps()
  end

  @doc false
  def changeset(record, attrs) do
    record
    |> cast(attrs, [:round_number, :time_limit_minutes, :status, :started_at, :ended_at, :tournament_id])
    |> validate_required([:round_number, :time_limit_minutes])
    |> validate_inclusion(:status, ["Pending", "Active", "Completed"])
    |> validate_number(:round_number, greater_than: 0, message: "Round number must be greater than zero")
    |> validate_number(:time_limit_minutes, greater_than: 0, message: "Round time limit must be greater than zero")
    |> then(fn cs ->
      if not is_nil(get_field(cs, :ended_at)) and (not ((get_field(cs, :ended_at) || 0) > get_field(cs, :started_at))) do
        Ecto.Changeset.add_error(cs, :ended_at, "Round end time must be after start time")
      else
        cs
      end
    end)
    |> then(fn cs ->
      if get_field(cs, :status) == "Completed" and (is_nil(get_field(cs, :started_at))) do
        Ecto.Changeset.add_error(cs, :started_at, "Completed round must have a start time")
      else
        cs
      end
    end)
  end

  # ── Business operations ────────────────────────────────────────────

  def start(_record) do
    # TODO: implement TournamentRound.start
    :ok
  end

  def complete(_record) do
    # TODO: implement TournamentRound.complete
    :ok
  end

  def generate_pairings(_record) do
    # TODO: implement TournamentRound.generate_pairings
    :ok
  end

  def is_time_expired(_record) do
    # TODO: implement TournamentRound.is_time_expired
    {:error, :not_implemented}
  end
end
